lib/core/src/Cardano/Wallet/Api/Link.hs

```diff
@@ -0,0 +1,520 @@
+    endpoint @(Api.JoinStakePool Net) (\mk -> mk sid wid)
+  where
+    sid = s ^. typed @(ApiT PoolId)
+    wid = w ^. typed @(ApiT WalletId)
+
+quitStakePool
+    :: forall (style :: WalletStyle) s w.
+        ( HasType (ApiT PoolId) s
+        , HasType (ApiT WalletId) w
+        , style ~ 'Shelley
+        )
+    => s
+    -> w
+    -> (Method, Text)
+quitStakePool s w =
+    endpoint @(Api.QuitStakePool Net) (\mk -> mk sid wid)
+  where
+    sid = s ^. typed @(ApiT PoolId)
+    wid = w ^. typed @(ApiT WalletId)
+
+delegationFee
+    :: forall (style :: WalletStyle) w.
+        ( HasType (ApiT WalletId) w
+        , style ~ 'Shelley
+        )
+    => w
+    -> (Method, Text)
+delegationFee w =
+    endpoint @Api.DelegationFee (wid &)
+  where
+    wid = w ^. typed @(ApiT WalletId)
+
+--
+-- Network Information
+--
+
+getNetworkInfo
+    :: (Method, Text)
+getNetworkInfo =
+    endpoint @Api.Network id
+
+--
+-- Proxy
+--
+
+postExternalTransaction
+    :: (Method, Text)
+postExternalTransaction =
+    endpoint @Api.PostExternalTransaction id
+
+
+--
+-- Internals
+--
+
+-- | A safe endpoint creator. This extracts the endpoint from a given Api type
+-- in the form of:
+--
+-- - A 'Text' string (the URL)
+-- - An HTTP 'Method' for calling this endpoint (e.g. GET, POST, ...)
+--
+-- This function does not type-check if the given endpoint (via type
+-- application) is not actually part of the wallet API.
+--
+-- Note that the 'MkLink endpoint Text' depends on the type of 'endpoint'.
+--
+-- - For simple endpoints with no path nor query parameters we have:
+--
+--   @MkLink endpoint Text ~ Text@ and therefore, we can simply do
+--
+--   >>> endpoint @Api.ListWallets id
+--   ( "GET", "/v2/wallets" )
+--
+-- - For endpoints with parameters, 'MkLink endpoint Text' will be a function
+--   taking as many arguments as there are path or query parameters (path
+--   parameters going first).
+--
+--   >>> endpoint @Api.GetWallet (\(mk :: ApiT WalletId -> Text) -> mk wid)
+--   ( "GET", "/v2/wallets/2512a00e9653fe49a44a5886202e24d77eeb998f" )
+--
+--   Or, simply:
+--
+--   >>> endpoint @Api.GetWallet (wid &)
+--   ( "GET", "/v2/wallets/2512a00e9653fe49a44a5886202e24d77eeb998f" )
+endpoint
+    :: forall endpoint.
+        ( IsElem  endpoint (Api Net)
+        , HasLink endpoint
+        , HasVerb endpoint
+        )
+    => (MkLink endpoint Text -> Text)
+    -> (Method, Text)
+endpoint mk =
+    ( method (Proxy @endpoint)
+    , "/v2" <> mk (safeLink' toUrlPiece (Proxy @(Api Net)) (Proxy @endpoint))
+    )
+
+-- Returns first argument for Shelley style wallet, second argument otherwise.
+discriminate
+    :: forall (style :: WalletStyle) a.
+        ( KnownSymbol (WalletStyleSymbol style)
+        )
+    => a
+    -> a
+    -> a
+discriminate shelley byron =
+    if isShelley
+        then shelley
+        else byron
+  where
+    isShelley = isJust $ sameSymbol
+        (Proxy @(WalletStyleSymbol 'Shelley))
+        (Proxy @(WalletStyleSymbol style))
```

Comment:
It seems this function would continue to compile if we added another `WalletStyle`.

But if that new wallet style is neither Shelly nor Byron, then we'd always end up returning the `byron` value.

Can this function be rewritten so that it pattern matches on all of the cases, and fails with a compiler error if a new wallet style is added in the future?

Reply:
> and fails with a compiler error

You like challenging me, don't you. I'll see what's possible to do here.